### ISSUE TITLE
Memory leak in pp_lexer_get_next_group_of_tokens_of_label()

### ISSUE THREAD

**Issue by ampli** [opencog/link-grammar#55]
In the function below, the memory for **tokens** is not getting freed by the callers.
There is a **free(tokens)** in the function, but it is a noop since tokens is always NULL there.
The memory leak is small, but can become larger if the en dict is reopened in a loop.

Possible fixes:
1. Add free(tokens) in the 7 places in which this functioned is invoked.
2. Add a **token** field in PPLexTable_s (initialized to NULL), and free it when PPLexTable_s is freed. In addition, use realloc() here. The function can also be changed to return **n_tokens**.

```
const char **pp_lexer_get_next_group_of_tokens_of_label(PPLexTable *lt, size_t *n_tokens)
{ 
  /* all tokens until next comma, null-terminated */
  int n;
  pp_label_node *p;
  static const char **tokens = NULL;  
  static int extents=0;

  p=lt->current_node_of_active_label;	   
  for (n=0; p!=NULL && strcmp(p->str,","); n++, p=p->next) {}
  if (n>extents) {
     extents = n;
     free (tokens);
     tokens = (const char **) malloc (extents * sizeof(const char*));
  }
  
... 

  *n_tokens = n;
  return tokens;
}
```

**Comment by linas** [comment]
Heh.
1) Either fix would be ok w/ me.
2) If you really want something to do, you can try to figure out what the original lex and yacc sources should be. These were lost long ago.

3) I suppose you've noticed: there are now unit tests in the 'tests' directory.  They are run w/ the 'make check'.  I just ran dict-reopen through valgrind, and it does not leak for me (42 bytes stuill reachable, at the end...)

**Comment by ampli** [comment]
1. Ok.

2. If this is the code in which I added support for specifying a doublequote character, then it was quite readable for me.  Anyway, I can rewrite it if needed. This may be needed if you would like to add non-trivial new dict syntax. But I think that simple additions can be done even now (e.g. adding a way to specify numerical character codes, like \x0160).

3. It reads data/ru. The first leak is in reading **en/4.0.knowledge**.

The second leak is probably what you found.  It doesn't accumulate, but has an associated bug (a permanently cached path_found, which may be incorrect when further opening a dictionary after chdir() or dictionary_set_data_dir()). I will send you a fix using cache_path_found()  (for not using external global variables).

**Comment by linas** [comment]
I just pushed a fix for this to the 5.2-main branch
